### DISCUSSION TITLE
Limitations of Supabase

### DISCUSSION THREAD

**Discussion by asyncink** [supabase/supabase#6512] [Questions] [answered]
Hi guys and Supabase team!

I've been working with Supabase and Next for a few months building isomorphic e-commerce application. As far as I can tell now, it's not possible to build an entire application using only Supabase client and Next.

The blocker is that the Supabase client is only accessing the database from the client side — so I can not deal with things like placing orders, creating transactions or user notifications from client side. Even with anon key every user can generate notifications for himself and these notifications will be valid. To do so, user don’t even need to dig into source code, just open browser dev tools and find the request. Having this request user is able to repeat it with curl, changing the parameters as he likes. Same thing with placing orders, changing account balance, etc.

Next offers API Routes for this, so I can isolate all necessary logic on server side and user will not be able to bypass server side checks. But Supabase client is not working server side so it is not the solution.

And then it turns out that Supabase is not an alternative to Firebase, with which you can build application, but a replacement for database providers such as Atlas. Convenient way to get database up and running.

Am I wrong and missing something? And how do you solve this issues in your projects?

**Comment by joshgeller** [comment, marked as answer]
Check out this course, it might answer some of your questions:

https://egghead.io/courses/build-a-saas-product-with-next-js-supabase-and-stripe-61f2bc20

**Reply by asyncink** [reply]
Thank you, will check it out. And if in short, what solution to the described problems is proposed there?

**Reply by asyncink** [reply]
Thank you @joshgeller, looks like this Egghead course gives proper solutions for my questions. They use Supabase client in API Routes, so I was wrong that Supabase client is client-side only.

**Comment by liaujianjie** [comment]
I moved my project with from Firebase to Supabase and Next.js+Vercel. I can assure you that Supabase can do **more** than what Firebase Firestore and Firebase Realtime Database is able to do. :-)

—

> The blocker is that the Supabase client is only accessing the database from the client side — so I can not deal with things like placing orders, creating transactions or user notifications from client side. Even with anon key every user can generate notifications for himself and these notifications will be valid. To do so, user don’t even need to dig into source code, just open browser dev tools and find the request. Having this request user is able to repeat it with curl, changing the parameters as he likes. Same thing with placing orders, changing account balance, etc.

You can restrict what a client can do with RLS.

> But Supabase client is not working server side so it is not the solution.

The Supabase client can also be authenticated with the service role key to perform what you'd typically expect the Firebase admin SDK to be able to do.

**Reply by asyncink** [reply]
No, you didn't get the point. With enabled and properly configured RLS user will be able do anything with database that policies allow. Check out examples in the first message of this thread.

**Comment by steve-chavez** [comment]
> Even with anon key 
> Having this request user is able to repeat it with curl, changing the parameters as he likes. Same thing with placing orders, changing account balance, etc.

Hm, why do you think RLS policies can't block what `anon` does? You can do:

```sql
create policy "profiles aren't viewable by anonymous users"
on profiles for select
to anon
using (
  false
);

-- the same can be done for all other tables and for authenticated users policies you can change `to anon` to `to authenticated`
```
